armory/utils/metrics.py

```diff
@@ -151,6 +157,44 @@ def _snr_spectrogram(x_i, x_adv_i):
     return signal_power / noise_power
 
 
+@contextmanager
+def resource_context(name="Name", profiler=None):
+    profiler_types = ["Basic", "Deterministic"]
+    if profiler is not None and profiler not in profiler_types:
+        raise ValueError(f"Profiler {profiler} is not one of {profiler_types}.")
+    if profiler == "Deterministic":
+        logger.warn(
+            "Using Deterministic profiler. This may reduce timing accuracy and result in a large results file."
+        )
+        pr = cProfile.Profile()
+        pr.enable()
+    if profiler == "Basic" or profiler == "Deterministic":
+        startTime = time.perf_counter()
+    yield
+    if profiler == "Basic" or profiler == "Deterministic":
+        elapsedTime = time.perf_counter() - startTime
+        if profiler == "Deterministic":
+            pr.disable()
+            s = io.StringIO()
+            sortby = "cumulative"
+            ps = pstats.Stats(pr, stream=s).sort_stats(sortby)
+            ps.print_stats()
+            stats = s.getvalue()
+        if name in computational_resource_dict:
+            computational_resource_dict[name]["execution_count"] += 1
+            computational_resource_dict[name]["total_time"] += elapsedTime
+            if profiler == "Deterministic":
+                computational_resource_dict[name]["stats"] += stats
+
+        else:
+            computational_resource_dict[name] = {}
+            computational_resource_dict[name]["execution_count"] = 1
+            computational_resource_dict[name]["total_time"] = elapsedTime
+            if profiler == "Deterministic":
+                computational_resource_dict[name]["stats"] = stats
```

Comment:
There's a lot of duplicate code here that could be collapsed. A defaultdict would work well for that, I think.

Something like:
```
from collections import defaultdict
...
if name not in computational_resource_dict:
    computational_resource_dict[name] = defaultdict(lambda: 0)  # uninitialized values default to 0
comp = computational_resource_dict[name]
comp["execution_count"] += 1
comp["total_time"] += elapsedTime
if profile == "Deterministic":
    comp["stats"] += stats
```